src/providers/baremetal.rs

```diff
@@ -0,0 +1,221 @@
+/// each instance of Setup corresponds to a single machine.
+#[derive(Clone)]
+pub struct Setup {
+    addr: std::net::SocketAddr,
+    username: String,
+    key_path: Option<std::path::PathBuf>,
+    setup_fn:
+        Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl super::MachineSetup for Setup {
+    type Region = String;
+    fn region(&self) -> Self::Region {
+        format!("bare:{}", self.addr)
+    }
+}
+
+impl PartialEq for Setup {
+    fn eq(&self, other: &Self) -> bool {
+        self.addr == other.addr
+    }
+}
+
+impl Eq for Setup {}
+
+impl std::hash::Hash for Setup {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.addr.hash(state);
+    }
+}
+
+impl Setup {
+    pub fn new(
+        addr: impl std::net::ToSocketAddrs,
+        username: Option<String>,
+    ) -> Result<Self, Error> {
+        let username: Result<String, Error> = username.map(Ok).unwrap_or_else(|| {
+            let user = String::from_utf8(std::process::Command::new("whoami").output()?.stdout)?;
+            let user = user
+                .split_whitespace()
+                .next()
+                .expect("expect newline after whoami output");
+            Ok(user.to_string())
+        });
+        let username = username?;
+
+        Ok(Self {
+            username,
+            addr: addr
+                .to_socket_addrs()?
+                .next()
+                .ok_or_else(|| format_err!("No socket addresses found"))?,
+            key_path: None,
+            setup_fn: None,
+        })
+    }
+
+    pub fn key_path(self, p: impl AsRef<std::path::Path>) -> Self {
+        Self {
+            key_path: Some(p.as_ref().to_path_buf()),
+            ..self
+        }
+    }
+
+    pub fn setup(
+        self,
+        setup: impl Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync + 'static,
+    ) -> Self {
+        Self {
+            setup_fn: Some(Arc::new(setup)),
+            ..self
+        }
+    }
+}
+
+/// Only one machine is supported per instance of this Launcher, further instances of `Setup`
+/// passed to `launch` will
+/// be ignored, since it doesn't make sense to connect to the same machine twice.
+///
+/// The `impl Drop` of this type is a no-op, since Tsunami can't terminate an existing machine.
+#[derive(Default)]
+pub struct Machine {
+    pub log: Option<slog::Logger>,
+    name: String,
+    addr: Option<std::net::SocketAddr>,
+    username: String,
+    key_path: Option<std::path::PathBuf>,
+}
+
+impl super::Launcher for Machine {
+    type Machine = Setup;
+
+    fn launch(&mut self, l: super::LaunchDescriptor<Self::Machine>) -> Result<(), Error> {
+        self.log = Some(l.log);
+        let log = self.log.as_ref().expect("Baremetal machine uninitialized");
+        let dscs = l
+            .machines
+            .into_iter()
+            .collect::<Vec<(String, Self::Machine)>>();
+        if dscs.is_empty() {
```

Comment:
I would make this an `assert!`. Trying to launch something that has no machines should be an error in `TsunamiBuilder` already I think.

Reply:
actually, I think `TsunamiBuilder` will happily just not launch anything and return `Ok(())` if you ask it to launch zero machines.
either way, good idea to make this an assert.

Reply:
actually, in the case that this type is called without going through `TsunamiBuilder`, it would retain more flexibility to not panic on this error? So maybe it is better to keep as is.